### PR TITLE
Fix #4570, #3824: Change all JS related code to use WKContentWorld sandboxing Part #1

### LINKED ISSUE #4570: Modernize evaluateJavascript, drop  <iOS 14 code for it
@Brandon-T  has few ideas how to make the evaluate javascript calls more robust. We should also remove the old code that makes the whole procedure more complex.

### PR DESCRIPTION
**Pull request by Brandon-T**
## Security Review
- https://github.com/brave/security/issues/692

## Summary of Changes
- Change all JS related code to use WKContentWorld sandboxing Part #1
- Fixed JS function generation code which incorrectly handled arrays, dictionaries, and primitive types. The previous code would NOT generate a proper Javascript array or dictionary. It also would convert all primitive types to `String`, so booleans and other primitives like Int, Float, Double, etc.. would NOT work without this PR.
- In Part-2, we need to go through every single script and see what is the strictest level of sandboxing we can use WITHOUT breaking the scripts.

This pull request fixes #4570, #3824

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:
- Test readermode
- Test Playlists
- Test Youtube Fullscreen
- Test WebAuthN
- Test Brave Search
- Test Brave Ads
- Test Find-In-Page
- Test Downloading
- Test blocking cookies

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Client/Frontend/Browser/New Tab Page/NewTabPageViewController.swift [hunk 1]
```diff
@@ -166,7 +166,9 @@ class NewTabPageViewController: UIViewController {
         }
         
         background.changed = { [weak self] in
-            self?.setupBackgroundImage()
+            DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
```
Comment (thypon) [inline Client/Frontend/Browser/New Tab Page/NewTabPageViewController.swift]: (nosec) Is this change related to WkContentWorld or it is just a quick fix for an old bug?

File: Client/Frontend/Browser/SessionRestoreHandler.swift [hunk 2]
```diff
@@ -13,9 +13,7 @@ extension WKWebView {
     // Use JS to redirect the page without adding a history entry
     func replaceLocation(with url: URL) {
         let safeUrl = url.absoluteString.replacingOccurrences(of: "'", with: apostropheEncoded)
-        evaluateSafeJavaScript(functionName: "location.replace", args: [safeUrl], sandboxed: true, escapeArgs: false, asFunction: true, completion: nil)
-        
-        evaluateJavaScript("location.replace('\(safeUrl)')", in: nil, in: .defaultClient, completionHandler: nil) //swiftlint:disable:this safe_javascript
```
Comment (thypon) [inline Client/Frontend/Browser/SessionRestoreHandler.swift:18]: (nosec) Is this not required anymore?

File: Client/Frontend/Browser/UserScriptManager.swift [hunk 3]
```diff
@@ -146,10 +146,16 @@ class UserScriptManager {
                 let wrappedSource = "(function() { const SECURITY_TOKEN = '\(UserScriptManager.messageHandlerTokenString)'; \(source) })()"
 
                 if sandboxed {
-                    return WKUserScript.createInDefaultContentWorld(source: wrappedSource, injectionTime: injectionTime, forMainFrameOnly: mainFrameOnly)
-                } else {
-                    return WKUserScript(source: wrappedSource, injectionTime: injectionTime, forMainFrameOnly: mainFrameOnly)
+                    return WKUserScript(source: wrappedSource,
+                                        injectionTime: injectionTime,
+                                        forMainFrameOnly: mainFrameOnly,
+                                        in: .defaultClient)
```
Comment (thypon) [inline Client/Frontend/Browser/UserScriptManager.swift:142]: is the `.defaultClient` JS context shared across rewards and all the user-scripts in sandboxed mode?

File: Client/Frontend/Browser/SessionRestoreHandler.swift [hunk 4]
```diff
@@ -13,9 +13,7 @@ extension WKWebView {
...
+        evaluateSafeJavaScript(functionName: "location.replace", args: [safeUrl], contentWorld: .defaultClient, escapeArgs: false, asFunction: true, completion: nil)
```
Comment (thypon) [inline Client/Frontend/Browser/SessionRestoreHandler.swift:16]: Was `escapeArgs` breaking some code path?

File: Shared/Extensions/WKWebViewExtensions.swift [hunk 5]
```diff
@@ -11,46 +11,44 @@ enum JavascriptError: Error {
...
-            context?.exceptionHandler = { context, exception in
-                if exception != nil {
-                    error = JavascriptError.invalid
+    func generateJSFunctionString(functionName: String, args: [Any?], escapeArgs: Bool = true) -> (javascript: String, error: Error?) {
+        var sanitizedArgs = [String]()
+        for arg in args {
+            if let arg = arg {
+                do {
+                    if let arg = arg as? String {
+                        sanitizedArgs.append(escapeArgs ? "'\(arg.htmlEntityEncodedString)'" : "\(arg)")
```
Comment (thypon) [inline Shared/Extensions/WKWebViewExtensions.swift:21]: Isn't it enough to use the JSONDecoder instead of the custom htmlEntityEncodedString?

https://stackoverflow.com/questions/59236554/how-to-use-swifts-jsonserialization-when-json-returns-a-single-string

File: Client/Frontend/Browser/SessionRestoreHandler.swift (hunk 2 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/SessionRestoreHandler.swift:18]: Not needed because the line above it already invokes it. For some reason there was duplicate code that would invoke it twice.

File: Client/Frontend/Browser/SessionRestoreHandler.swift (hunk 4 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/SessionRestoreHandler.swift:16]: Yes it would break it. Also the original code has `escapeArgs: false` and I didn't want to change anything. The `escapeArgs` would wrap this into a string and it is already a string, so it will escape it by escaping the quotes which causes it not to work anymore.

File: Client/Frontend/Browser/UserScriptManager.swift (hunk 3 above)
Comment (Brandon-T) [reply Client/Frontend/Browser/UserScriptManager.swift:142]: Yes it is. It is shared with all our injected scripts that use `.defaultClient`. If we want a script to be completely separate from any others, we need to have that script have its own content world.

File: Shared/Extensions/WKWebViewExtensions.swift (hunk 5 above)
Comment (Brandon-T) [reply Shared/Extensions/WKWebViewExtensions.swift:21]: The original code was using the `htmlEntityEncodedString` I believe by itself, the `JSONDecoder` or `JSONEncoder` doesn't escape `'` character. Only `"`.

**Review by thypon — APPROVED** [review]

File: DataTests/CoreDataTestCase.swift [hunk 6]
```diff
@@ -15,6 +16,7 @@ class CoreDataTestCase: XCTestCase {
                                                name: NSNotification.Name.NSManagedObjectContextDidSave,
                                                object: nil)
         
+        Preferences.Database.DocumentToSupportDirectoryMigration.completed.value = true
```
Comment (kylehickinson) [inline DataTests/CoreDataTestCase.swift:19]: Unrelated?

File: Shared/Extensions/WKWebViewExtensions.swift [hunk 7]
```diff
@@ -11,46 +11,44 @@ enum JavascriptError: Error {
...
-
-        args.forEach {
-            if !escapeArgs {
-                sanitizedArgs.append("\($0)")
-                return
-            }
-            context?.exceptionHandler = { context, exception in
-                if exception != nil {
-                    error = JavascriptError.invalid
+    func generateJSFunctionString(functionName: String, args: [Any?], escapeArgs: Bool = true) -> (javascript: String, error: Error?) {
```
Comment (kylehickinson) [inline Shared/Extensions/WKWebViewExtensions.swift:15]: Probably better off returning `Result<String, Error>`

**Comment by diracdeltas** [comment]
Note that the associated security review was closed. If there are any further changes to this PR that may have security risk, please feel free to reopen, thanks

**Review by iccub — APPROVED** [review]
approved but please take a look at Kyle's comments

**Comment by iccub** [comment]
Closing, will be handled here https://github.com/brave/brave-ios/pull/4859
